tsl/src/continuous_aggs/refresh.c

```diff
@@ -78,50 +78,93 @@ get_largest_bucketed_window(Oid timetype, int64 bucket_width)
+compute_bucketed_refresh_window(const InternalTimeRange *const refresh_window,
+								const int64 bucket_width, const bool shrink)
 {
 	InternalTimeRange result = *refresh_window;
 	InternalTimeRange largest_bucketed_window =
 		get_largest_bucketed_window(refresh_window->type, bucket_width);
 
-	if (result.start <= largest_bucketed_window.start)
+	if (refresh_window->start <= largest_bucketed_window.start)
 		result.start = largest_bucketed_window.start;
 	else
-		result.start = ts_time_bucket_by_type(bucket_width, result.start, result.type);
+	{
+		if (shrink)
+		{
+			/* The start time needs to be aligned with the first fully enclosed bucket.
+			 * So the original window start is moved to next bucket, except if the start is
+			 * already aligned to a bucket, thus 1 is substracted to stay in the curret
+			 * bucket in the aligned case. */
+			int64 included_bucket = ts_time_saturating_add(refresh_window->start,
+														   bucket_width - 1,
+														   refresh_window->type);
+			/* Get the start of the included bucket. */
+			result.start =
+				ts_time_bucket_by_type(bucket_width, included_bucket, refresh_window->type);
+		}
+		else
+		{
+			/* For alignment with bucket, which includes the start of the refresh window, we just
+			 * need to get start of the bucket. */
+			result.start =
+				ts_time_bucket_by_type(bucket_width, refresh_window->start, refresh_window->type);
+		}
+	}
 
 	if (result.end >= largest_bucketed_window.end)
 		result.end = largest_bucketed_window.end;
 	else
 	{
-		int64 exclusive_end = result.end;
-		int64 bucketed_end;
-
-		/* The end of the window is non-inclusive so subtract one before
-		 * bucketing in case we're already at the end of the bucket (we don't
-		 * want to add an extra bucket). But we also don't want to subtract if
-		 * we are at the start of the bucket (we don't want to remove a
-		 * bucket). The last  */
-		if (result.end > result.start)
-			exclusive_end = ts_time_saturating_sub(result.end, 1, result.type);
-
-		bucketed_end = ts_time_bucket_by_type(bucket_width, exclusive_end, result.type);
-
-		/* We get the time value for the start of the bucket, so need to add
-		 * bucket_width to get the end of it */
-		result.end = ts_time_saturating_add(bucketed_end, bucket_width, refresh_window->type);
+		if (shrink)
+		{
+			/* The window is reduced to the beginning of the bucket, which contains the end of
+			 * the refresh window. */
+			result.end =
+				ts_time_bucket_by_type(bucket_width, refresh_window->end, refresh_window->type);
+		}
+		else
+		{
+			int64 exclusive_end = result.end;
+			int64 bucketed_end;
+
+			/* The end of the window is non-inclusive so subtract one before
+			 * bucketing in case we're already at the end of the bucket (we don't
+			 * want to add an extra bucket). But we also don't want to subtract if
+			 * we are at the start of the bucket (we don't want to remove a
+			 * bucket). */
+			if (result.end > result.start)
```

Comment:
If the end of the window is exclusive, shouldn't it always be the case that `result.end > result.start`?

Reply:
This was originally there and I didn't modify the code. However, I agree that the false condition is not possible. cc @erimatnor

Reply:
Replaced with assert.